csrc/multidevice/utils.cpp

```diff
@@ -151,6 +217,116 @@ void insertReshardings(Fusion* fusion) {
+    NVF_ERROR(
+        shard_additions.size() + shard_deletions.size() <= 1,
+        "Resharding expr can only support one axis")
+
+    // For gather operations i.e. ID goes from sharded to unsharded
+    // this will rematerialize a sharded axis.
+    // ProcessGroup expects contiguous tensors.
+    // Update input to push the rematerialized axis to the front -> collective
+    // -> permute the rematerizlied axis to the proper location Example: [i0
+    // DIDx(i1)] -> [i0 i1] Rewritten to: [DIDx(i1) i0] -> [i1 i0] -> [i0 i1]
+    // Note: there are no reduction based collectives that materializes an axis
+    // so expr is guaranteed to be a set.
+    if (!shard_deletions.empty()) {
+      IterDomain* shard_deleted_id = shard_deletions[0];
+      int sharding_axis =
+          static_cast<int>(input->domain()->rootPosOf(shard_deleted_id));
+      if (isOutermostAllocatedId(input, shard_deleted_id)) {
+        continue;
+      }
+      TensorView* input_permute = permute(input, {{sharding_axis, 0}});
+      TensorView* output_permute = set(input_permute);
+      TensorView* new_output = permute(output_permute, {{0, sharding_axis}});
+      ir_utils::replaceValInAllExprInputsAndFusionOutputs(output, new_output);
+
+      // Propagate shardings from input and manually apply sharding deletions.
+      shardAllLike(input, {input_permute, output_permute, new_output});
+      output_permute->axis(0)->parallelize(ParallelType::Serial);
+      new_output->axis(sharding_axis)->parallelize(ParallelType::Serial);
+      output_permute->setDeviceMesh(output->getDeviceMesh());
+      new_output->setDeviceMesh(output->getDeviceMesh());
+    }
+    // For scatter operations i.e. ID goes from unsharded to sharded
+    // Update input to push the scattered axis to the front -> collective ->
+    // permute the sharded axis to the proper location.
+    // Example: [i0 DIDx(i1) i2] -> [i0 r1 DIDx(i2)]
+    // Rewritten to: [i0 DIDx(i1) i2] -> [i2 i0 DIDx(i1)] ->
+    //                    [DIDx(i2) i0 r1] -> [i0 DIDx(i2)]
+    // Note that reduction axis shifts from axis=1 to axis=2.
+    else if (!shard_additions.empty()) {
+      auto shard_added_id = shard_additions[0];
+      int sharding_axis =
+          static_cast<int>(output->domain()->rootPosOf(shard_added_id));
+      if (isOutermostAllocatedId(output, shard_added_id)) {
+        continue;
+      }
+
+      TensorView* input_permute = permute(input, {{sharding_axis, 0}});
+      TensorView* output_permute = nullptr;
+      // Calculate the number of reduction axes before the sharding axis.
+      // After permuting the sharding axis to the front, the reduction axis
+      // will be offset by this amount.
+      auto reduction_axis = output->getReductionAxis();
+      int num_reduction_axes_before_sharding_axis =
+          (reduction_axis.has_value() &&
+           sharding_axis > static_cast<int>(reduction_axis.value()))
+          ? 1
```

Comment:
I guess we implicitely assume here that there can be maximum one reduction axis?
Can you please make this explicity by adding an assert ? see https://github.com/NVIDIA/Fuser/blob/9c2359fe3f7655a0e4726a06ba186ecb58a5c2e4/csrc/scheduler/mma_utils.cpp#L1444